packages/astro/src/runtime/server/render/astro/render.ts

```diff
@@ -165,3 +168,119 @@ async function bufferHeadContent(result: SSRResult) {
 		}
 	}
 }
+
+export async function renderToAsyncIterable(
+	result: SSRResult,
+	componentFactory: AstroComponentFactory,
+	props: any,
+	children: any,
+	isPage = false,
+	route?: RouteData
+): Promise<AsyncIterable<Uint8Array> | Response> {
+  const templateResult = await callComponentAsTemplateResultOrResponse(
+    result,
+    componentFactory,
+    props,
+    children,
+    route
+  );
+  if (templateResult instanceof Response)
+    return templateResult;
+  let renderedFirstPageChunk = false;
+  if (isPage) {
+    await bufferHeadContent(result);
+  }
+
+	// This implements the iterator protocol:
+	// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_async_iterator_and_async_iterable_protocols
+	// The `iterator` is passed to the Response as a stream-like thing.
+	// The `buffer` array acts like a buffer. During render the `destination` pushes
+	// chunks of Uint8Arrays into the buffer. The response calls `next()` and we combine
+	// all of the chunks into one Uint8Array and then empty it.
+	
+	let error: Error | null = null;
+	// The `next` is an object `{ promise, resolve, reject }` that we use to wait
+	// for chunks to be pushed into the buffer.
+  let next = promiseWithResolvers<void>();
+  const buffer: Uint8Array[] = []; // []Uint8Array
+
+  const iterator = {
+    async next() {
+      await next.promise;
+
+			// If an error occurs during rendering, throw the error as we cannot proceed.
+			if(error) {
+				throw error;
+			}
+
+      // Get the total length of all arrays.
+      let length = 0;
+			for(let i = 0, len = buffer.length; i < len; i++) {
+				length += buffer[i].length;
+			}
+
+      // Create a new array with total length and merge all source arrays.
+      let mergedArray = new Uint8Array(length);
```

Comment:
Did you consider using Buffer.concat? In Bun, we've optimized Buffer.concat a lot, it should be meaningfully faster than this approach.

If for some reason `Buffer.concat` cannot be used, `Buffer.allocUnsafe` should also be faster here - there's no risk of uninitialized memory leaking since the full length is counted and will not change

Reply:
No I did not know that it would be faster. Would be happy to take a PR on this, we have benchmarks we can compare.